### PR TITLE
WT-7566 Fix modify after free memory error for dead dhandles.

### PR DESCRIPTION
**Pull request by ddanderson**

### CODE REVIEW

**Comment by ddanderson** [comment]
@keithbostic, do you have a chance to review this?  I know you've looked at the cursor caching code before.  There's a lot of background info in https://jira.mongodb.org/browse/BF-21150 .  If you don't have the cycles, let me know.  Thanks!

File: src/include/cursor_inline.h [hunk 1]
```diff
@@ -352,10 +352,14 @@ __wt_cursor_dhandle_decr_use(WT_SESSION_IMPL *session)
...
-    /* If we close a handle with a time of death set, clear it. */
+    /*
+     * If we close a handle with a time of death set, clear it. The ordering is important: after
+     * decrementing the use count, there's a chance that the data handle can be freed.
+     */
     WT_ASSERT(session, dhandle->session_inuse > 0);
-    if (__wt_atomic_subi32(&dhandle->session_inuse, 1) == 0 && dhandle->timeofdeath != 0)
+    if (dhandle->timeofdeath != 0 && dhandle->session_inuse == 1)
         dhandle->timeofdeath = 0;
+    (void)__wt_atomic_subi32(&dhandle->session_inuse, 1);
```
Comment (keithbostic) [inline src/include/cursor_inline.h:362]: Can you confirm we don't get into trouble if we set `timeofdeath` to 0, but then `session_inuse` isn't decremented to 0, that is, another thread increments `session_inuse` between our setting `timeofdeath` and our decrement of `session_inuse`?

I took a look and don't see any problems, but wanted to confirm with you.

**Review by keithbostic — APPROVED** [review]
lgtm

File: src/include/cursor_inline.h (hunk 1 above)
Comment (ddanderson) [reply src/include/cursor_inline.h:362]: I don't think this is a problem.  Even if it's not perfect with an occasional race (so it's set or not set inappropriately) it doesn't much matter.  Setting it to 0 communicates to the sweep thread - hey someone just used this handle, don't mark it dead yet.  So it has heuristic value only.

**Comment by ddanderson** [comment]
evergreen retry

**Pull request merged**